Kernel/Graphics/VirtIOGPU/GraphicsAdapter.cpp

```diff
@@ -374,41 +392,43 @@ void VirtIOGraphicsAdapter::flush_displayed_image(Graphics::VirtIOGPU::ResourceI
     request.resource_id = resource_id.value();
     request.rect = dirty_rect;
 
-    synchronous_virtio_gpu_command(start_of_scratch_space(), sizeof(request), sizeof(response));
+    TRY(synchronous_virtio_gpu_command(100, start_of_scratch_space(), sizeof(request), sizeof(response)));
 
-    VERIFY(response.type == to_underlying(Graphics::VirtIOGPU::Protocol::CommandType::VIRTIO_GPU_RESP_OK_NODATA));
+    if (response.type == to_underlying(Graphics::VirtIOGPU::Protocol::CommandType::VIRTIO_GPU_RESP_OK_NODATA))
+        return {};
+    return EIO;
 }
 
-void VirtIOGraphicsAdapter::synchronous_virtio_gpu_command(PhysicalAddress buffer_start, size_t request_size, size_t response_size)
+ErrorOr<void> VirtIOGraphicsAdapter::synchronous_virtio_gpu_command(size_t milliseconds_timeout, PhysicalAddress buffer_start, size_t request_size, size_t response_size)
 {
     VERIFY(m_operation_lock.is_locked());
-    VERIFY(m_outstanding_request.is_empty());
     auto& queue = get_queue(CONTROLQ);
-    {
-        SpinlockLocker lock(queue.lock());
-        VirtIO::QueueChain chain { queue };
-        chain.add_buffer_to_chain(buffer_start, request_size, VirtIO::BufferType::DeviceReadable);
-        chain.add_buffer_to_chain(buffer_start.offset(request_size), response_size, VirtIO::BufferType::DeviceWritable);
-        supply_chain_and_notify(CONTROLQ, chain);
-        full_memory_barrier();
+    queue.disable_interrupts();
+    SpinlockLocker lock(queue.lock());
+    VirtIO::QueueChain chain { queue };
+    chain.add_buffer_to_chain(buffer_start, request_size, VirtIO::BufferType::DeviceReadable);
+    chain.add_buffer_to_chain(buffer_start.offset(request_size), response_size, VirtIO::BufferType::DeviceWritable);
+    supply_chain_and_notify(CONTROLQ, chain);
+    full_memory_barrier();
+    size_t current_time = 0;
+    ScopeGuard clear_used_buffers([&] {
+        queue.discard_used_buffers();
+    });
+    while (current_time < milliseconds_timeout) {
+        if (queue.new_data_available())
```

Comment:
Is it possible to have new data available here immediately or do we always hit the microseconds delay?

Reply:
I just did a test and it looks like we are always hitting the 1ms delay. When I change this to an increasing delay, starting at 1 microsecond VirGLDemo runs super smooth.

Reply:
Maybe something for a followup PR tho

Reply:
I will keep this unresolved so we can take care of this in a future PR. Until then it's fine this way :)

Reply:
On second thought, I decided to change this to microseconds granularity, because it seems that it makes everything much more responsive :)